[Fix] Change RactiveChooser to reference options using their indices not values (Fixes #444)

omargfh opened the pull request:
## Relevant Issue
Fixes #444

## Steps to Remedy Issue
1. Set the `value` attribute on `<option>` in `RactiveChooser` to use the `index` position of the choice rather than its value.
2. Bound the `<select>` to `widget.currentChoice` instead of the widget's `internalValue`.
3. Removed the `widget.currentValue` observer responsible for setting the `widget.currentChoice` (index) value on the widget controller.
4. Added a new local event `chooser-option-change` and set it to update the `internalValue` and `widget.currentChoice` based on the new selected index before firing `widget-value-change` to the `WidgetController`.

## Consequences and Side Effects
RactiveChooser refers to options using their indices not values, making each value unique even when congruent or equivalent to other values.**

> ** Note that if you have choices `[1, 2, 3, 1]` and have the _**last**_ `1` selected then append a new choice to the list, say, `[1, 2, 3, 4, 1]`, the event system will change your choice from the _**last**_ `1` to the _**first**_ `1` in the current implementation. This applies to options that match in _value_ and _type_. Congruent values (but different types) do not exhibit this behavior.

## Tested Environments

- [x] Chrome Canary on MacOS 15.5 (ARM)
- [x] Firefox on MacOS 15.5 (ARM)
- [ ] Chromium on Ubuntu 24 LTS (ARM)
- [ ] Firefox on Ubuntu 24 LTS (ARM)
- [ ] Chromium on Windows 10/11 (ARM)
- [ ] Firefox on Windows 10/11 (ARM)
- [ ] Chromium on Windows 10/11 (64-bit)
- [ ] Firefox on Windows 10/11 (64-bit)

omargfh committed:
Bugfix(RactiveChooser): fix #444 Change Chooser to index-based choices over value-based choices to deter ambiguity when congruent options are present

LaCuneta commented on app/assets/javascripts/beak/widgets/ractives/chooser.coffee:
```diff
@@ -154,12 +154,18 @@ RactiveChooser = RactiveValueWidget.extend({
+        @set('internalValue', selectedChoice)
+        @set('widget.currentChoice', if selectedIdx >= 0 then selectedIdx else 0)
+        @fire('widget-value-change')
+        return
```
Can either drop the `return` here or also `return false`.

LaCuneta commented on app/assets/javascripts/beak/widgets/ractives/chooser.coffee:
```diff
@@ -154,12 +154,18 @@ RactiveChooser = RactiveValueWidget.extend({
+    'chooser-option-change': (event) ->
+      widget       = @get('widget')
+      selectedIdx  = parseInt(event.node.value)
+      selectedChoice = widget.choices[selectedIdx] or widget.choices[0]
```
It won't cause an error, but it seems odd to me to get the `selectedChoice` before checking if the index is valid.  I think it could be moved inside the `if isSelectedIdxValid` conditional?  That might also remove the need to mix in the `or widget.choices[0]`.

LaCuneta commented on app/assets/javascripts/beak/widgets/ractives/chooser.coffee:
```diff
@@ -154,12 +154,18 @@ RactiveChooser = RactiveValueWidget.extend({
-      return
+  on: {
+    'chooser-option-change': (event) ->
+      widget       = @get('widget')
```
I do prefer the assignments to be whitespace-aligned to the `=`, but it should be for all assignments in this block.  Having a standard 1 space before the `=` is fine, too, as long as it's consistent.

LaCuneta reviewed (changes requested):
Looking good, thank you!  Using the indices instead of the values looks like the right approach to me.  One issue I noticed is that with the loss of the `widget.currentValue` observer, the chooser no longer updates when I change its value through code instead of the UI.  If you load Wolf Sheep Predation it should default to a `model-version` with `"sheep-wolves-grass"` as the value.  If you run `set model-version "sheep-wolves"` in the Command Center, the widget will not update, but the backing global will.

I also test #373 with the new code, and it resolves that older issue as well.

I also left a few notes in the code.  Let me know if you have questions on anything.

omargfh committed:
Bugfix(RactiveChooser): Chooser does not reflect changes from engine

LaCuneta reviewed (changes requested):
Thanks for the updates!  I left a few more code comments on small things.

The update to add in an unexpected value to the choices `<select>` is clever!  But this is a weird situation where the engine (Tortoise) should be rejecting that unexpected choice as an error (as NetLogo desktop does).  This is a weird situation as some widget global variables don't follow this "out of band error" rule (sliders).  For now, I think setting the chosen index to `-1` so the select shows blank is fine, until we get the error properly thrown in Tortoise sometime later on.

omargfh committed:
Minor (Chooser, #docs): reset graceful handling for invalid currentChoice

Values:
- `internalChoice`: Tracker for the internal value of the chooser,
                    can be -1 if no valid choice is selected
- `internalValue`:  Used by the ValueWidget to track updates to the value
                    and requires @fire('widget-value-changed') to notify
                    the outside world of changes.
- `widget.currentValue`: The current value of the chooser, can be invalid
                        (out of bounds index) and comes from the model
- `widget.currentChoice`: Unused in Chooser but it is the chooser
                          responsibility to keep it in sync with the model.
                          This will be (0) if the value is invalid and
                          (0, ..., n-1) if valid, where n is the
                          number of choices.

Current Lifecycle:
- Chooser is created with a valid internalChoice index (0)

- @initialization: Chooser syncs up the internal values and model value
  with the `widget.currentChoice` from the model.
  - If this is a new widget, this will be (0)

- @user-input: Chooser listens to user input, tracking the selected index
  and syncs up all other values accordingly. It also notifies the outside
  world of the change via `widget-value-changed` event.

  P.S. This will cause a dupe call as the ValueWidget will update the
  model value and the model will notify the widget of the change.
  We dedupe this by checking if the value actually changed.

- @model-update: Chooser listens to changes in the model value and
  updates the internal values accordingly. This excludes `internalValue`
  as the ValueWidget tracks it.

- @options-change: When the list of options changes, the chooser detects
  the change by subscribing to the root '*.refresh-choices'. This will
  only update the index of the current choice.

Behavior of dupe choices:
- Dupe choices of the same type and different types are allowed.
- Only behavior of dupes of different types is defined and is defined
  as distinct choices.
- Behavior of dupes of the same type is undefined but will not brick the
  chooser, model, session, or browser. Currently, choice order is main-
  -tained on user input but not on model update. This means that if you
  choose (1, [1], 2) and then change the model value to (1), the
  chooser may select either the first (1) or the second (1) depending
  on the implementation. Current implementation selects the first (1).

Invalid choice handling:
- Invalid choices show as blank in the chooser (internalChoice = -1).
- Invalid choices set the `internalValue` to undefined.
- Invalid choices set the `widget.currentChoice` to (0) to keep the
  model in a valid state. This will reset the chooser to a valid option
  once model is reloaded.
- Invalid choices do not affect the `widget.currentValue` which
  can be invalid. Tortoise will still run with invalid choices.

The @_setChoice method:
- Dedupes duplicate value updates by checking if the value
  actually changed.
- Handles invalid values according to the above spec.
- Updates all internal values accordingly, with `internalValue`
  being optional.
- Does NOT notify the outside world of changes. This is
  handled by explicit calls to `@fire('widget-value-changed')`.

Other behavior related to invalid values and dupes is undefined.

LaCuneta reviewed (approved):
Thanks for the updates!  I tested it out and things are working as I'd expect now.  I had a few comments in the code for review, but I'm just going to merge and make some updates so I can keep working towards a release.  Just let me know if you have questions on anything.
